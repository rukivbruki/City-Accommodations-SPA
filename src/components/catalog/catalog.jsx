import React, {Fragment} from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import PlacesList from "../places-list/places-list";
import CatalogEmpty from "../catalog-empty/catalog-empty";
import Tabs from "../tabs/tabs";
import Sort from "../sort/sort";
import withActiveElement from "../../hocs/with-active-element/with-active-element";
import Map from "../map/map";
import {getActiveCity, getCoordinates, getHoveredOffer, getSortedOffers} from "../../reducer/data/selector";
import witSelectState from "../../hocs/with-select-state/with-select-state";

const TabsWrapped = withActiveElement(Tabs);
const SortWrapped = witSelectState(Sort);

const Catalog = (props) => {
  const {
    currentCity,
    offers,
    mapCoordinates,
    activeOffer,
  } = props;

  return (
    <Fragment>
      {offers.length ?
        <main className="page__main page__main--index">
          <h1 className="visually-hidden">Cities</h1>

          <TabsWrapped currentCity={currentCity}/>

          <div className="cities">
            <div className="cities__places-container container">
              <section className="cities__places places">
                <h2 className="visually-hidden">Places</h2>
                <b className="places__found">{offers.length} places to stay in {currentCity}</b>
                <form className="places__sorting" action="#" method="get">
                  <span className="places__sorting-caption">Sort by </span>

                  <SortWrapped/>

                </form>

                <PlacesList offers={offers}/>

              </section>
              <div className="cities__right-section">
                <section className="cities__map">
                  <Map coordinates={mapCoordinates} currentCity={currentCity} activeOffer={activeOffer}/>
                </section>
              </div>
            </div>
          </div>
        </main>
        : <CatalogEmpty currentCity={currentCity}/>
      }
    </Fragment>
  );
};

Catalog.propTypes = {
  activeOffer: PropTypes.shape({
    id: PropTypes.number.isRequired,
    city: PropTypes.shape({
      name: PropTypes.string.isRequired,
      location: PropTypes.arrayOf(PropTypes.number).isRequired,
      zoom: PropTypes.number.isRequired
    }),
    location: PropTypes.shape({
      coordinates: PropTypes.arrayOf(PropTypes.number).isRequired,
      zoom: PropTypes.number.isRequired
    }),
    title: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    previewImage: PropTypes.string.isRequired,
    rating: PropTypes.number.isRequired,
    isPremium: PropTypes.bool.isRequired,
    isFavorite: PropTypes.bool.isRequired,
  }),
  currentCity: PropTypes.string,
  mapCoordinates: PropTypes.array.isRequired,
  offers: PropTypes.array,
};

const mapStateToProps = (state) => ({
  activeOffer: getHoveredOffer(state),
  currentCity: getActiveCity(state),
  mapCoordinates: getCoordinates(state),
  offers: getSortedOffers(state),
});

export {Catalog};
export default connect(mapStateToProps)(Catalog);
