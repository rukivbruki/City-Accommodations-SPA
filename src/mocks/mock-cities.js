export default [
  {
    name: `City name 1`,
    location: {
      latitude: 1,
      longitude: 2,
      zoom: 13
    }
  },
  {
    name: `City name 2`,
    location: {
      latitude: 2,
      longitude: 3,
      zoom: 13
    }
  },
];
