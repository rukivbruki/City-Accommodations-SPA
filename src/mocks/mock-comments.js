export default {
  id: 0,
  isLoading: false,
  loadingError: null,
  isLoaded: true,
  comments: [
    {
      id: 0,
      user: {
        id: 0,
        isPro: true,
        name: `name`,
        avatarUrl: `url`
      },
      rating: 4,
      comment: `some text`,
      date: `2019-05-14T06:07:35.291Z`
    },
    {
      id: 1,
      user: {
        id: 0,
        isPro: true,
        name: `name 2`,
        avatarUrl: `url 2`
      },
      rating: 4,
      comment: `some text 2`,
      date: `2018-05-14T06:07:35.291Z`
    },
  ]
};
