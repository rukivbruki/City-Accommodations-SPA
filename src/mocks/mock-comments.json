[
  {
    "id": 1,
    "user": {
      "id": 14,
      "is_pro": true,
      "name": "Corey",
      "avatar_url": "url"
    },
    "rating": 5,
    "comment": "text",
    "date": "2019-05-16T00:39:04.525Z"
  },
  {
    "id": 2,
    "user": {
      "id": 19,
      "is_pro": false,
      "name": "Christina",
      "avatar_url": "url"
    },
    "rating": 2,
    "comment": "text",
    "date": "2019-05-16T00:39:04.525Z"
  }
]
