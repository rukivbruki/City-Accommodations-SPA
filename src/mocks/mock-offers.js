
export default [
  {
    id: 999,
    isPremium: true,
    picture: `test-image1.jpg`,
    price: 111,
    rating: 22,
    title: `Test title 1`,
    description: `text`,
    bedrooms: 2,
    type: `apartment`,
    goods: [`one`, `two`, `tree`],
    images: [`url1`, `url2`, `url3`],
    previewImage: ``,
    maxAdults: 2,
    isFavorite: true,
    city: {
      name: `City name 1`,
      location: {
        latitude: 1,
        longitude: 2,
        zoom: 13
      }
    },
    location: {
      latitude: 1,
      longitude: 2,
      zoom: 13
    },
    host: {
      id: 1,
      name: `name`,
      isPro: true,
      avatarUrl: `url`
    }
  },
  {
    id: 222,
    isPremium: false,
    picture: `test-image2.jpg`,
    price: 222,
    rating: 100,
    title: `Test title 2`,
    description: `text`,
    goods: [`one`, `two`, `tree`],
    images: [`url1`, `url2`, `url3`],
    bedrooms: 3,
    maxAdults: 1,
    type: `house`,
    previewImage: ``,
    isFavorite: true,
    coordinates: [52.3909553943508, 4.85309666406198],
    city: {
      name: `City name 2`,
      location: {
        latitude: 1,
        longitude: 2,
        zoom: 13
      }
    },
    location: {
      latitude: 1,
      longitude: 2,
      zoom: 13
    },
    host: {
      id: 1,
      name: `name`,
      isPro: true,
      avatarUrl: `url`
    }
  },
];
