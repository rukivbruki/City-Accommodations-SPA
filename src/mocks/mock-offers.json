[
  {
    "city": {
      "name": "City 1",
      "location": {
        "latitude": 1,
        "longitude": 2,
        "zoom": 13
      }
    },
    "preview_image": "https://test.com/1.jpg",
    "images": [
      "https://test.com/1.jpg"
    ],
    "title": "Text",
    "is_favorite": false,
    "is_premium": false,
    "rating": 5,
    "type": "house",
    "bedrooms": 3,
    "max_adults": 7,
    "price": 100,
    "goods": [
      "Text",
      "Text"
    ],
    "host": {
      "id": 25,
      "name": "Angelina",
      "is_pro": true,
      "avatar_url": "img/avatar.jpg"
    },
    "description": "Text",
    "location": {
      "latitude": 2,
      "longitude": 3,
      "zoom": 16
    },
    "id": 1
  },
  {
    "city": {
      "name": "City 2",
      "location": {
        "latitude": 1,
        "longitude": 2,
        "zoom": 13
      }
    },
    "preview_image": "https://test.com/1.jpg",
    "images": [
      "https://test.com/1.jpg"
    ],
    "title": "Text",
    "is_favorite": false,
    "is_premium": false,
    "rating": 5,
    "type": "house",
    "bedrooms": 3,
    "max_adults": 7,
    "price": 100,
    "goods": [
      "Text",
      "Text"
    ],
    "host": {
      "id": 25,
      "name": "Angelina",
      "is_pro": true,
      "avatar_url": "img/avatar.jpg"
    },
    "description": "Text",
    "location": {
      "latitude": 2,
      "longitude": 3,
      "zoom": 16
    },
    "id": 2
  }
]
