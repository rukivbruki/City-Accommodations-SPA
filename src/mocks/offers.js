export const OFFERS = [
  {
    id: 1,
    city: `Hamburg`,
    title: `Amazing and Extremely Central Flatn`,
    previewImage: `https://es31-server.appspot.com/six-cities/static/hotel/19.jpg`,
    isPremium: false,
    isFavorite: false,
    price: 103,
    rating: 4.3,
    type: `room`,
    coordinates: [53.570341000000006, 9.975654]
  },
  {
    id: 2,
    city: `Paris`,
    title: `Canal View Prinsengracht`,
    previewImage: `https://es31-server.appspot.com/six-cities/static/hotel/4.jpg`,
    price: 109,
    isPremium: false,
    isFavorite: false,
    rating: 4,
    type: `room`,
    coordinates: [48.846610000000005, 2.374499]
  },
  {
    id: 3,
    city: `Amsterdam`,
    title: `Canal View Prinsengracht`,
    previewImage: `https://es31-server.appspot.com/six-cities/static/hotel/11.jpg`,
    price: 754,
    isPremium: false,
    isFavorite: true,
    rating: 3.4,
    type: `house`,
    coordinates: [52.35754, 4.9179759999999995]
  },
  {
    id: 4,
    city: `Amsterdam`,
    title: `The Joshua Tree House`,
    previewImage: `https://es31-server.appspot.com/six-cities/static/hotel/4.jpg`,
    price: 210,
    isPremium: true,
    isFavorite: false,
    rating: 4,
    type: `room`,
    coordinates: [52.36554, 4.911976]
  },
  {
    id: 5,
    city: `Paris`,
    title: `Penthouse, 4-5 rooms + 5 balconies`,
    previewImage: `https://es31-server.appspot.com/six-cities/static/hotel/19.jpg`,
    price: 855,
    isPremium: true,
    isFavorite: false,
    rating: 3.6,
    type: `house`,
    coordinates: [48.83861, 2.350499]
  },
  {
    id: 6,
    city: `Brussels`,
    title: `Nice, cozy, warm big bed apartment`,
    previewImage: `https://es31-server.appspot.com/six-cities/static/hotel/14.jpg`,
    price: 371,
    isPremium: false,
    isFavorite: true,
    rating: 4.1,
    type: `house`,
    coordinates: [50.835556999999994, 4.354697]
  },
  {
    id: 7,
    city: `Dusseldorf`,
    title: `Nice, cozy, warm big bed apartment`,
    previewImage: `https://es31-server.appspot.com/six-cities/static/hotel/3.jpg`,
    price: 149,
    isPremium: true,
    isFavorite: true,
    rating: 4.8,
    type: `house`,
    coordinates: [51.225402, 6.784314]
  },
  {
    id: 8,
    city: `Cologne`,
    title: `Canal View Prinsengracht`,
    previewImage: `https://es31-server.appspot.com/six-cities/static/hotel/20.jpg`,
    price: 765,
    isPremium: false,
    isFavorite: true,
    rating: 4.8,
    type: `apartment`,
    coordinates: [50.951361, 6.944974]
  },
  {
    id: 9,
    city: `Hamburg`,
    title: `The house among olive `,
    previewImage: `https://es31-server.appspot.com/six-cities/static/hotel/13.jpg`,
    price: 786,
    isPremium: false,
    isFavorite: true,
    rating: 3.4,
    type: `house`,
    coordinates: [53.573341000000006, 9.994654]
  },
  {
    id: 10,
    city: `Brussels`,
    title: `Loft Studio in the Central Area`,
    previewImage: `https://es31-server.appspot.com/six-cities/static/hotel/8.jpg`,
    price: 400,
    isPremium: true,
    isFavorite: true,
    rating: 2,
    type: `room`,
    coordinates: [50.822556999999996, 4.347697]
  }
];
