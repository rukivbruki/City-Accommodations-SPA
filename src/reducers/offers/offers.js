import {ActionCreator as ActiveCityActionCreator} from '../active-city/active-city';
import parseOffer from '../../utils/parse-offer';
import {getActiveCity} from '../active-city/selectors';

const CITIES_COUNT = 6;

const initialState = {
  isLoaded: false,
  loadingError: null,
  offers: null,
  cities: null,
};

const ActionType = {
  OFFERS_LOAD: `OFFERS_LOAD`,
  OFFERS_LOADING_ERROR: `OFFERS_LOADING_ERROR`,
  OFFERS_LOADED: `OFFERS_LOADED`,
  OFFERS_UPDATE_OFFER: `OFFERS_UPDATE_OFFER`,
};

const ActionCreator = {
  loadOffers: () => ({
    type: ActionType.OFFERS_LOAD,
  }),

  loadingOffersError: (error) => ({
    type: ActionType.OFFERS_LOADING_ERROR,
    payload: error,
  }),

  loadedOffers: (offers, cities) => ({
    type: ActionType.OFFERS_LOADED,
    payload: {offers, cities}
  }),

  updateOffer: (offer) => ({
    type: ActionType.OFFERS_UPDATE_OFFER,
    payload: offer,
  }),
};

const getCitiesFromOffers = (offers) => {
  const cities = [];
  const map = new Map();
  for (const offer of offers) {
    if (!map.has(offer.city.name)) {
      map.set(offer.city.name, true);
      cities.push(offer.city);
    }
  }
  return cities.slice(0, CITIES_COUNT);
};

const loadOffers = () => (dispatch, getState, api) => {
  dispatch(ActionCreator.loadOffers());
  return api.get(`/hotels`)
    .then((response) => {
      const offers = response.data.map((it) => parseOffer(it));
      const cities = getCitiesFromOffers(offers);
      if (!getActiveCity(getState())) {
        const activeCity = cities[Math.floor(Math.random() * cities.length)];
        dispatch(ActiveCityActionCreator.setActiveCity(activeCity));
      }
      dispatch(ActionCreator.loadedOffers(offers, cities));
    })
    .catch((error) => {
      let errorString = `Loading comments error :(`;
      if (error.response && error.response.data && error.response.data.error) {
        errorString = error.response.data.error;
      }
      dispatch(ActionCreator.loadingOffersError(errorString));
    });
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.OFFERS_LOAD:
      return Object.assign({}, state, {
        isLoaded: false,
        loadingError: null,
      });

    case ActionType.OFFERS_LOADING_ERROR:
      return Object.assign({}, state, {
        isLoaded: false,
        loadingError: action.payload,
      });

    case ActionType.OFFERS_LOADED:
      return Object.assign({}, state, {
        isLoaded: true,
        loadingError: null,
        offers: action.payload.offers,
        cities: action.payload.cities,
      });

    case ActionType.OFFERS_UPDATE_OFFER:
      return Object.assign({}, state, {
        offers: state.offers.map((offer) => {
          if (offer.id === action.payload.id) {
            return action.payload;
          }
          return offer;
        })
      });
  }

  return state;
};

export {
  getCitiesFromOffers,
  ActionCreator,
  ActionType,
  loadOffers,
  reducer,
};
