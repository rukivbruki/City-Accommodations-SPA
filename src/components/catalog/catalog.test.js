import React from "react";
import renderer from "react-test-renderer";
import {Catalog} from "./catalog";
import {OFFERS} from "../../mocks/offers";

jest.mock(`../map/map.jsx`, () => jest.fn().mockReturnValue(null));
jest.mock(`../tabs/tabs.jsx`, () => jest.fn().mockReturnValue(null));
jest.mock(`../sort/sort.jsx`, () => jest.fn().mockReturnValue(null));
jest.mock(`../places-list/places-list.jsx`, () => jest.fn().mockReturnValue(null));

const citiesCoordinates = [[52.123123, 4.8123123], [52.21313, 4.9123123]];

it(`Catalog component render correct`, () => {
  const catalog = renderer
    .create(
        <Catalog
          citiesCoordinates={citiesCoordinates}
          currentCity={`NY`}
          offers={OFFERS}
          mapCoordinates={[]}
        />)
    .toJSON();

  expect(catalog).toMatchSnapshot();
});
