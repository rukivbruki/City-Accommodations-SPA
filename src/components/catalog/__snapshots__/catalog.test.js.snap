// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Catalog component render correct 1`] = `
<main
  className="page__main page__main--index"
>
  <h1
    className="visually-hidden"
  >
    Cities
  </h1>
  <div
    className="cities"
  >
    <div
      className="cities__places-container container"
    >
      <section
        className="cities__places places"
      >
        <h2
          className="visually-hidden"
        >
          Places
        </h2>
        <b
          className="places__found"
        >
          10
           places to stay in 
          NY
        </b>
        <form
          action="#"
          className="places__sorting"
          method="get"
        >
          <span
            className="places__sorting-caption"
          >
            Sort by 
          </span>
        </form>
      </section>
      <div
        className="cities__right-section"
      >
        <section
          className="cities__map"
        />
      </div>
    </div>
  </div>
</main>
`;
